src/wallet.rs

```diff
@@ -0,0 +1,962 @@
+                bytes = decrypt(&bytes, pwd)?;
+                if bytes.len() != SEED_SIZE {
+                    return Err(Error::WalletFileCorrupted);
+                }
+                seed.copy_from_slice(&bytes);
+            }
+            "2.0" => {
+                let content = decrypt(&bytes, pwd)?;
+                // extract seed
+                seed.copy_from_slice(&content[0..SEED_SIZE]);
+                // extract addrs
+                addrs = Addresses::from_bytes(content[SEED_SIZE..].to_vec())?
+            }
+            _ => {
+                return Err(Error::UnknownFileVersion);
+            }
+        };
+
+        // create and return
+        Ok(Self {
+            wallet: None,
+            addrs,
+            store: LocalStore::new(seed),
+            file: Some(file),
+            status: |_| {},
+        })
+    }
+
+    /// Attempts to load a legacy wallet file (no version number)
+    fn from_legacy_file(file: F) -> Result<Self, Error> {
+        let path = file.path();
+        let pwd = file.pwd();
+
+        // attempt to load and decode wallet
+        let mut bytes = fs::read(&path.0)?;
+
+        // check for old version information and strip it if present
+        if bytes[1] == 0 && bytes[2] == 0 {
+            bytes.drain(0..3);
+        }
+
+        bytes = decrypt(&bytes, pwd)?;
+        if bytes.len() != SEED_SIZE {
+            return Err(Error::WalletFileCorrupted);
+        }
+
+        // get our seed
+        let mut seed = [0u8; SEED_SIZE];
+        seed.copy_from_slice(&bytes);
+
+        // return the store
+        Ok(Self {
+            wallet: None,
+            addrs: Addresses::default(),
+            store: LocalStore::new(seed),
+            file: Some(file),
+            status: |_| {},
+        })
+    }
+
+    /// Saves wallet to file from which it was loaded
+    pub fn save(&self) -> Result<(), Error> {
+        match &self.file {
+            Some(f) => {
+                // create file content
+                let seed = self.store.get_seed()?;
+                let mut content = seed.to_vec();
+                content.append(&mut self.addrs.to_vec());
+
+                // encrypt everything
+                content = encrypt(&content, f.pwd())?;
+
+                // prepend magic number and encoding version information
+                let mut prefix = [0u8; MAGIC.len() + VERSION.len()];
+                prefix[..MAGIC.len()].copy_from_slice(MAGIC);
+                prefix[MAGIC.len()..].copy_from_slice(VERSION);
+                content.splice(0..0, prefix.iter().cloned());
+
+                // write to file
+                fs::write(Path::new(&f.path().0), content)?;
+                Ok(())
+            }
+            None => Err(Error::WalletFileMissing),
+        }
+    }
+
+    /// Saves wallet to the provided file
+    pub fn save_to(&mut self, file: F) -> Result<(), Error> {
+        // set our new file
+        self.file = Some(file);
+        self.save()
+    }
```

Comment:
I tried to do it but the compiler didn't like it. Let's go over it together.

Reply:
We agreed to leave this as is and update the docs accordingly to be clear on the fact that `save_to` changes the internal file and any subsequent call to `save` will use that file.

Keeping the file is necessary because a user that has created or recovered a wallet will get an error as soon as he/she generates a new address, as we'll try to save to a non-existing file.